File: templates/roadmap/aiml/learning/04-core-fundamentals/nodejs/01-core.js
/**
 * app.js — COMPLETE JavaScript CORE (Working)
 * No frameworks. No servers. No libraries.
 */

console.log("===== JS CORE START =====");

/* ===============================
1. VARIABLES & DATA TYPES
================================ */
var legacy = "var";
let name = "Alice";
const age = 30;
let active = true;
let nothing = null;
let undef;
let big = 12345678901234567890n;

console.log(name, age, active, nothing, undef, big);

/* ===============================
2. TYPE CHECKING
================================ */
console.log(typeof name, typeof big, Array.isArray([]));

/* ===============================
3. STRINGS
================================ */
let text = " Hello World ";
console.log(text.trim().toUpperCase());
console.log(`User: ${name}, Age: ${age}`);

/* ===============================
4. OPERATORS
================================ */
let total = 10 + 5;
let power = 2 ** 3;
let status = total > 10 ? "High" : "Low";
console.log(total, power, status);

/* ===============================
5. CONTROL FLOW
================================ */
if (age >= 18) {
  console.log("Adult");
} else {
  console.log("Minor");
}

for (let i = 0; i < 3; i++) {
  console.log("For:", i);
}

/* ===============================
6. FUNCTIONS
================================ */
function greet(user) {
  return "Hello " + user;
}

const add = (a, b) => a + b;

console.log(greet("Bob"), add(2, 3));

/* ===============================
7. ARRAYS
================================ */
let numbers = [1, 2, 3];
numbers.push(4);

let doubled = numbers.map(n => n * 2);
let filtered = numbers.filter(n => n > 2);
let sum = numbers.reduce((a, b) => a + b, 0);

console.log(numbers, doubled, filtered, sum);

/* ===============================
8. OBJECTS
================================ */
let user = {
  id: 1,
  name: "Alice",
  greet() {
    return "Hi " + this.name;
  }
};

user.city = "Delhi";
delete user.id;

console.log(user, user.greet());

/* ===============================
9. DESTRUCTURING
================================ */
const { name: username } = user;
console.log(username);

/* ===============================
10. SET & MAP
================================ */
const set = new Set([1, 2, 2, 3]);
set.add(4);

const map = new Map();
map.set("a", 1);

console.log([...set], map.get("a"));

/* ===============================
11. DATE & MATH
================================ */
const now = new Date();
console.log(now.toISOString(), Math.sqrt(16));

/* ===============================
12. JSON
================================ */
const json = JSON.stringify(user);
const parsed = JSON.parse(json);
console.log(parsed);

/* ===============================
13. ERROR HANDLING
================================ */
try {
  throw new Error("Test error");
} catch (e) {
  console.log("Caught:", e.message);
} finally {
  console.log("Finally block executed");
}

/* ===============================
14. FILE SYSTEM
================================ */
const fs = require("fs");

fs.writeFileSync("data.txt", "Hello JS");
const fileData = fs.readFileSync("data.txt", "utf8");
console.log("File:", fileData);

/* ===============================
15. CLASSES & OOP
================================ */
class Person {
  constructor(name) {
    this.name = name;
  }
  greet() {
    return `Hi ${this.name}`;
  }
}

class Employee extends Person {}

const emp = new Employee("Charlie");
console.log(emp.greet());

/* ===============================
16. PROMISE & ASYNC
================================ */
function asyncTask() {
  return new Promise(resolve =>
    setTimeout(() => resolve("Async Done"), 500)
  );
}

async function runAsync() {
  const result = await asyncTask();
  console.log(result);
}

runAsync().then(() => {
  console.log("===== JS CORE END =====");
});

File: templates/roadmap/aiml/learning/04-core-fundamentals/python/01-core.py
"""
app.py — COMPLETE Python CORE (Working)
No frameworks. No web. No external libs.
"""

print("===== PYTHON CORE START =====")

# ===============================
# 1. VARIABLES & DATA TYPES
# ===============================
name = "Alice"
age = 30
active = True
nothing = None
big = 12345678901234567890

print(name, age, active, nothing, big)

# ===============================
# 2. TYPE CHECKING
# ===============================
print(type(name), isinstance(age, int))

# ===============================
# 3. STRINGS
# ===============================
text = " Hello World "
print(text.strip().upper())
print(f"User: {name}, Age: {age}")

# ===============================
# 4. OPERATORS
# ===============================
total = 10 + 5
power = 2 ** 3
status = "High" if total > 10 else "Low"
print(total, power, status)

# ===============================
# 5. CONTROL FLOW
# ===============================
if age >= 18:
    print("Adult")
else:
    print("Minor")

for i in range(3):
    print("For:", i)

# ===============================
# 6. FUNCTIONS
# ===============================
def greet(user):
    return "Hello " + user

add = lambda a, b: a + b

print(greet("Bob"), add(2, 3))

# ===============================
# 7. LISTS
# ===============================
numbers = [1, 2, 3]
numbers.append(4)

doubled = [n * 2 for n in numbers]
filtered = [n for n in numbers if n > 2]
total_sum = sum(numbers)

print(numbers, doubled, filtered, total_sum)

# ===============================
# 8. DICTIONARIES
# ===============================
user = {
    "id": 1,
    "name": "Alice"
}

user["city"] = "Delhi"
del user["id"]

print(user)

# ===============================
# 9. UNPACKING
# ===============================
username = user["name"]
print(username)

# ===============================
# 10. SET & TUPLE
# ===============================
unique = {1, 2, 2, 3}
unique.add(4)

point = (10, 20)
x, y = point

print(unique, x, y)

# ===============================
# 11. DATE & MATH
# ===============================
from datetime import datetime
import math

now = datetime.now()
print(now.isoformat(), math.sqrt(16))

# ===============================
# 12. JSON
# ===============================
import json

json_data = json.dumps(user)
parsed = json.loads(json_data)
print(parsed)

# ===============================
# 13. ERROR HANDLING
# ===============================
try:
    raise Exception("Test error")
except Exception as e:
    print("Caught:", e)
finally:
    print("Finally block executed")

# ===============================
# 14. FILE HANDLING
# ===============================
with open("data.txt", "w") as f:
    f.write("Hello Python")

with open("data.txt") as f:
    print("File:", f.read())

# ===============================
# 15. CLASSES & OOP
# ===============================
class Person:
    def __init__(self, name):
        self.name = name

    def greet(self):
        return f"Hi {self.name}"

class Employee(Person):
    pass

emp = Employee("Charlie")
print(emp.greet())

# ===============================
# 16. GENERATOR
# ===============================
def generator():
    yield 1
    yield 2

for val in generator():
    print("Yield:", val)

# ===============================
# 17. THREADING (BASIC)
# ===============================
import threading
import time

def task():
    time.sleep(0.5)
    print("Thread done")

t = threading.Thread(target=task)
t.start()
t.join()

print("===== PYTHON CORE END =====")
